---
title: "Preprocessing UKB Multi-omic data"
output:
  word_document: default
  html_notebook: default
editor_options:
  markdown:
    wrap: 72
---

```{r}
source("../config.R") #Loads libraries, variables and global function

```

# Preparing import commands

```{r}
#### Prepare vector for reading out correct metabolite columns (once done and added to fread function in import section, this has no longer to be run until new metabolites come in)
#NMR_Metabolomics_Index \<- read_excel("\~/PostDoc/Results/Metabolomics/NMR MetabolomicsIndex.xlsx")
#NMR_Metabolomics_Index$datafield <- paste0('"', NMR_Metabolomics_Index$datafield,'-0.0",') #nmrstring \<- cat(NMR_Metabolomics_Index\$datafield, sep="", file="metabolomics_index.txt")

#### Prepare vector for reading out serum parameter columns (once done and added to fread function in import section, this has no longer to be run until new metabolites come in)

#Blood_Marker_Index \<- read_excel("C:/Users/Jan/OneDrive - rwth-aachen.de/Dokumente/PostDoc/Results/Serum Marker/Blood parameters.xlsx")
#Blood_Marker_Index$datafield <- paste0('"', Blood_Marker_Index$datafield, '-0.0",') 
#setwd("\~/PostDoc/Results/Serum Marker") 
#Blood_Marker_String \<- cat( Blood_Marker_Index\$datafield, sep="", file="blood_marker_index.txt") #when pasting to fread function, delete last comma, and add )) at end of command for syntax correction
```


# Loading Data:

Importing and preprocessing of multimodal data

| Nr  | Description      | Dataset         |        n |
|-----|------------------|:----------------|---------:|
| 1a  | Patient IDs      | df_eid          |   502411 |
| 1b  | Covariates       | df_covariates   |  Varying |
| 1c  | ICD Diagnosis    | df_diagnosis    |   442277 |
| 2   | Blood parameters | df_blood        |   456855 |
| 3   | Metabolomics     | df_metabolomics | 248286   |
| 4   | Genomics         | df_snp          | 488149   |
| 5   | Radiomics        | df_mri          | \~50.000 | (Pending)


#### 1. EID, ICD, Covariates, PAR(Importing tables previously processed)
```{r}
#setwd(drive)
#df_eid <- fread("ukb52200.csv", select=c("eid")) %>%    #all eids in UKB
#write.csv(df_eid, file="C:/Users/Jan/OneDrive/Dokumente/PostDoc/Patient_tables/UKB_Patients_eids.csv", row.names=FALSE)    

#a) df_eid 
setwd(sharepoint_ukb)
df_withdrawals <- read.delim("raw/withdrawals.txt")
df_eid <- fread("extracted/UKB_Patients_eids.csv") %>% check_and_remove_withdrawals(df_withdrawals)

#b) Covariate data for all 502411 patients: df_covariates 502411
#Importing df_covariates: Preprocessing performed in Script "Extract and process covariates"
setwd(project_path)
load("data/dataframes/df_covariates.RData")
df_covariates <- check_and_remove_withdrawals(df_covariates, df_withdrawals)
#df_cap_covariates <- read_excel(paste0(sharepoint_ukb, "/", master_table), sheet="Cap_covariates") 


#c) Diagnosis data for selected  diagnosis (import diagnosis data for selected diagnosis in Script "Extract_multiple_Diagnosis"))
setwd(project_path)
load("data/dataframes/df_diagnosis.RData")
df_diagnosis <- df_diagnosis %>% check_and_remove_withdrawals(df_withdrawals)
#sanity(df_diagnosis)

setwd(sharepoint_ukb)
mapper <- read_excel(master_table, sheet="Mapper")

```

```{r}
# hist(df_covariates$Bloodpressure_sys)
# hist(df_covariates2$Bloodpressure_sys)
```

```{r}
# pre-check filtering of the dataframe for loss of DOI cases

# filter_rows_with_pos_entries(df_diagnosis)
# 
# innerjoin_df_y(filter_rows_with_pos_entries(df_diagnosis))

```
```{r}
# setwd(sharepoint_ukb)
# additional <- fread("raw/ukb52200.csv", select=c("eid", "30030-0.0"))

```

#### 2. Blood sample data for all 502411 patients df_blood 502411

```{r}
#setwd(sharepoint_ukb)
#import blood sample data from baseline and change names to measured parameters
#df_blood <- fread("raw/ukb52200.csv", select=c("eid", "30160-0.0","30220-0.0","30150-0.0","30210-0.0","30030-0.0","30020-0.0","30300-0.0","30290-0.0","30280-0.0","30120-0.0","30180-0.0","30050-0.0","30060-0.0","30040-0.0","30100-0.0","30260-0.0","30270-0.0","30130-0.0","30190-0.0","30140-0.0","30200-0.0","30170-0.0","30230-0.0","30080-0.0","30090-0.0","30110-0.0","30010-0.0","30070-0.0","30250-0.0","30240-0.0","30000-0.0","30620-0.0","30600-0.0","30610-0.0","30630-0.0","30640-0.0","30650-0.0","30710-0.0","30680-0.0","30690-0.0","30700-0.0","30720-0.0","30660-0.0","30730-0.0","30740-0.0","30750-0.0","30760-0.0","30770-0.0","30780-0.0","30790-0.0","30800-0.0","30810-0.0","30820-0.0","30830-0.0","30850-0.0","30840-0.0","30860-0.0","30870-0.0","30880-0.0","30670-0.0","30890-0.0"))
#setwd(sharepoint_ukb)
#write.csv(df_blood, file="extracted/UKB_Patients_blood.csv, row.names=FALSE)    

setwd(sharepoint_ukb)
df_blood <- fread("extracted/UKB_Patients_blood.csv") %>% check_and_remove_withdrawals(df_withdrawals)

# changing names to measured parameters
Blood_Marker_Index <- read_excel(master_table, sheet="Blood count and biochemistry") %>%
  mutate(datafield = paste0(datafield, '-0.0'))

df_blood <- df_blood %>%
  rename_with(
    ~ Blood_Marker_Index$Description[match(., Blood_Marker_Index$datafield)],
    .cols = -eid  # Exclude 'eid' from renaming
  )

df_blood <- df_blood %>%
  select(eid, all_of(Blood_Marker_Index$Description)) # Order Blood columns

#Overview and handling of NAs by imputing the most common value or removing
na_blood_col <- data.frame(na_columnwise(df_blood))
na_blood_row <- rowSums(is.na(df_blood))
hist(na_blood_row)
#write.xlsx(na_blood_col, file=paste(project, "/supplement/NA_count_df_blood.xlsx", sep=''))

if (na_mode == "impute") {
  summary(df_blood[,1:5])
  df_blood <- df_blood %>% 
    mutate_at(vars(-1), mean.impute) %>% # Apply mean.impute to all columns except the first one
    select(-c("Rheumatoid factor", "Oestradiol")) #delete columns with mostly NA (Oestradiol 402.879 NAs; Rheumatoid factor 437.202 NAs)
  summary(df_blood[,1:5])
  
} else if (na_mode == "remove") {     # Your removal code here
  df_blood <- omit.NA(df_blood, 20)   # Delete rows with >= 20 NAs per Row
  df_blood <- df_blood %>% #delete columns with mostly NA values (Lipoprotein A 111.652 NAs; Oestradiol 402.879 NAs; Rheumatoid factor 437.202 NAs)
  select(-c("Rheumatoid factor", "Oestradiol"))     
}
#save(df_blood, file=paste(project_path, "/data/dataframes/df_blood.RData", sep='')) # Save df_blood as .RData (preserves column classes)

vec_blood <- setdiff(colnames(df_blood), "eid") #this gets important for minmax normalization later on

#Sanity check
#sanity(df_blood)


# Select all patients with elevated liver enzymes at baseline
if (project_key == "hcc"){
    df_blood_risk <- df_blood %>%
  select(c("eid", "Aspartate aminotransferase", "Alanine aminotransferase", "Gamma glutamyltransferase" )) %>%
  inner_join(select(df_covariates, eid, SEX), by = "eid") %>%
  mutate(
    Elevated_AST = if_else(SEX == "Female" & `Aspartate aminotransferase` > 35 | SEX == "Male" & `Aspartate aminotransferase` > 50, 1, 0),
    Elevated_ALT = if_else(SEX == "Female" & `Alanine aminotransferase` > 35 | SEX == "Male" & `Alanine aminotransferase` > 50, 1, 0),
    Elevated_GGT = if_else(SEX == "Female" & `Gamma glutamyltransferase` > 40 | SEX == "Male" & `Gamma glutamyltransferase` > 60, 1, 0)
  ) %>%
  mutate(Elevated_Liver_Enzymes = if_else(Elevated_AST == 1 | Elevated_ALT == 1 | Elevated_GGT == 1, 1, 0)) %>%
  select(-c("Aspartate aminotransferase", "Alanine aminotransferase", "Gamma glutamyltransferase", "SEX"))
}


if (project_key == "cca") {   
  df_blood_risk <- df_blood %>%
  select(c("eid", "Aspartate aminotransferase", "Alanine aminotransferase", "Gamma glutamyltransferase", "Alkaline phosphatase" )) %>%
  inner_join(select(df_covariates, eid, SEX), by = "eid") %>%
  mutate(
    Elevated_AST = if_else(SEX == "Female" & `Aspartate aminotransferase` > 35 | SEX == "Male" & `Aspartate aminotransferase` > 50, 1, 0),
    Elevated_ALT = if_else(SEX == "Female" & `Alanine aminotransferase` > 35 | SEX == "Male" & `Alanine aminotransferase` > 50, 1, 0),
    Elevated_GGT = if_else(SEX == "Female" & `Gamma glutamyltransferase` > 40 | SEX == "Male" & `Gamma glutamyltransferase` > 60, 1, 0),
    Elevated_AP = if_else(SEX == "Female" & `Alkaline phosphatase` > 105 | SEX == "Male" & `Alkaline phosphatase` > 130, 1, 0)
  ) %>%
  mutate(Elevated_Liver_Enzymes = if_else(Elevated_AST == 1 | Elevated_ALT == 1 | Elevated_GGT == 1, 1, 0)) %>%
  select(-c("Aspartate aminotransferase", "Alanine aminotransferase", "Gamma glutamyltransferase", "SEX", "Alkaline phosphatase"))
  
}

vec_blood_risk <- colnames(select(df_blood_risk, -eid)) #pass into vector to later remove these columns

#df_diagnosis <- inner_join(df_blood_risk, df_diagnosis, by="eid")



#df_par <- filter_rows_with_pos_entries(df_diagnosis)
#innerjoin_df_y(df_test_par)


df_blood <- adjust_outliers(df_blood, c("HbA1c", "Basophill (%)"))




```

```{r}
hist(df_blood$`Basophill (%)`)
```



#### 3. Metabolomics data for 250k patients df_metabolomics

```{r}
#Import metabolomics (data for n=106804 patients, obtained from EDTA plasma from baseline recruitment)
#df_metabolomics<- fread("ukb674682.csv", select=c("eid", "23474-0.0","23475-0.0","23476-0.0","23477-0.0","23460-0.0","23479-0.0","23440-0.0","23439-0.0","23441-0.0","23433-0.0","23432-0.0","23431-0.0","23484-0.0","23526-0.0","23561-0.0","23533-0.0","23498-0.0","23568-0.0","23540-0.0","23505-0.0","23575-0.0","23547-0.0","23512-0.0","23554-0.0","23491-0.0","23519-0.0","23580-0.0","23610-0.0","23635-0.0","23615-0.0","23590-0.0","23640-0.0","23620-0.0","23595-0.0","23645-0.0","23625-0.0","23600-0.0","23630-0.0","23585-0.0","23605-0.0","23485-0.0","23418-0.0","23527-0.0","23417-0.0","23562-0.0","23534-0.0","23499-0.0","23569-0.0","23541-0.0","23506-0.0","23576-0.0","23548-0.0","23513-0.0","23416-0.0","23555-0.0","23492-0.0","23520-0.0","23581-0.0","23611-0.0","23636-0.0","23616-0.0","23591-0.0","23641-0.0","23621-0.0","23596-0.0","23646-0.0","23626-0.0","23601-0.0","23631-0.0","23586-0.0","23606-0.0","23473-0.0","23404-0.0","23481-0.0","23430-0.0","23523-0.0","23429-0.0","23558-0.0","23530-0.0","23495-0.0","23565-0.0","23537-0.0","23502-0.0","23572-0.0","23544-0.0","23509-0.0","23428-0.0","23551-0.0","23488-0.0","23516-0.0","23478-0.0","23443-0.0","23450-0.0","23457-0.0","23486-0.0","23422-0.0","23528-0.0","23421-0.0","23563-0.0","23535-0.0","23500-0.0","23570-0.0","23542-0.0","23507-0.0","23577-0.0","23549-0.0","23514-0.0","23420-0.0","23556-0.0","23493-0.0","23521-0.0","23582-0.0","23612-0.0","23637-0.0","23617-0.0","23592-0.0","23642-0.0","23622-0.0","23597-0.0","23647-0.0","23627-0.0","23602-0.0","23632-0.0","23587-0.0","23607-0.0","23470-0.0","23461-0.0","23462-0.0","23480-0.0","23406-0.0","23463-0.0","23465-0.0","23405-0.0","23471-0.0","23466-0.0","23449-0.0","23456-0.0","23447-0.0","23454-0.0","23444-0.0","23451-0.0","23445-0.0","23459-0.0","23452-0.0","23468-0.0","23437-0.0","23434-0.0","23483-0.0","23414-0.0","23525-0.0","23413-0.0","23560-0.0","23532-0.0","23497-0.0","23567-0.0","23539-0.0","23504-0.0","23574-0.0","23546-0.0","23511-0.0","23412-0.0","23553-0.0","23490-0.0","23518-0.0","23579-0.0","23609-0.0","23634-0.0","23614-0.0","23589-0.0","23639-0.0","23619-0.0","23594-0.0","23644-0.0","23624-0.0","23599-0.0","23629-0.0","23584-0.0","23604-0.0","23446-0.0","23458-0.0","23453-0.0","23472-0.0","23402-0.0","23448-0.0","23455-0.0","23438-0.0","23400-0.0","23401-0.0","23436-0.0","23464-0.0","23427-0.0","23415-0.0","23442-0.0","23419-0.0","23482-0.0","23426-0.0","23524-0.0","23425-0.0","23559-0.0","23531-0.0","23496-0.0","23423-0.0","23566-0.0","23538-0.0","23503-0.0","23573-0.0","23545-0.0","23510-0.0","23424-0.0","23552-0.0","23489-0.0","23517-0.0","23411-0.0","23407-0.0","23487-0.0","23410-0.0","23529-0.0","23409-0.0","23564-0.0","23536-0.0","23501-0.0","23571-0.0","23543-0.0","23508-0.0","23578-0.0","23550-0.0","23515-0.0","23408-0.0","23557-0.0","23494-0.0","23522-0.0","23435-0.0","23583-0.0","23613-0.0","23638-0.0","23618-0.0","23593-0.0","23643-0.0","23623-0.0","23598-0.0","23648-0.0","23628-0.0","23603-0.0","23633-0.0","23588-0.0","23608-0.0","23469-0.0","23403-0.0","23467-0.0"))
#write.csv(df_metabolomics, file="C:/Users/Jan/OneDrive/Dokumente/PostDoc/Patient_tables/UKB_Patients_metabolomics.csv", row.names=FALSE)    

setwd(sharepoint_ukb)
df_metabolomics <- fread("extracted/metabolomics250k.csv") %>% check_and_remove_withdrawals(df_withdrawals)
metabolomics_index <- read_excel(master_table, sheet="NMR_Metabolomics")

# Select non-Na participants only  
df_metabolomics <- na.omit(df_metabolomics)
df_metabolomics$V1 <- NULL #Remove unnecessary v1 column
nrow(df_metabolomics) #correlates to number of participants still included (should be 248286, 248266 after withdrawal removal)

#Subset for directly measured metabolites
df_metabolomics <-  setnames(df_metabolomics, old = metabolomics_index$datafield1, new = metabolomics_index$name_processing, skip_absent=TRUE)
summary(df_metabolomics[,1:5])
#Set names of metabolites
#Order the columns
col_order <- match(metabolomics_index$name_processing, colnames(df_metabolomics))
col_order <- c(1, col_order)
df_metabolomics <- as.data.frame(df_metabolomics)
df_metabolomics <- df_metabolomics[,col_order]

df_metabolomics <- subset(df_metabolomics[1:144]) 

vec_metabolomics <- setdiff(colnames(df_metabolomics), "eid") #this gets important for minmax normalization later on

#sanity(df_metabolomics)
```

#### 4. Genomics

```{r}
# Set up vector with all chromosomes where there are RAW files
# create "empty" dataframe with eids to later store SNPs
# Collect info from RAW files in for loop and build df with all snps
# duplicate df with merging on "eid" (This creates a _x and _y column for every SNP)
# rename _x and _y in _hom and _het and change scale to 0-1



#Set up a folder inside your project folder called genetics. Inside this folder you store the RAW files for the chromosomes. The following lines of code should put the numberof the chromosomes for which there are Raw files into the vector vec

setwd(project_path)
df_snp <- df_eid
files <- list.files("genetics")
vec <- sort(
  unlist(
    lapply(
      strsplit(files, "_"),
      function(x) as.numeric(gsub("[^0-9]", "", x)))))
df_snp <- df_eid  #create "empty" dataframe with eids to later store SNPs

setwd(sharepoint_ukb)
for (i in vec) {  #Loop through each SNP
  print(i)  #control if loop is working
  snp <- data.frame(read.csv(paste(project_path, "/genetics/", DOI, "_snps_chr",i, ".raw", sep=""), sep=""))  %>% #import snp, change this name according to your file names!
    rename(c("eid"="FID")) %>%  #eid should always be called eid for merging etc...
    dplyr::select(-c("IID", "PAT", "MAT", "SEX", "PHENOTYPE"))  # only eid and rsXYZ are relevant later on.
  df_snp <- merge(df_snp, snp, by.x="eid", by.y="eid", all.x=TRUE)  #append snps to df
}

# Convert the numeric values to factors with desired levels
df_snp[,2:ncol(df_snp)] <- lapply(df_snp[,2:ncol(df_snp)], function(column) {
  factor(column, levels=c(0, 1, 2), labels=c("wt", "het", "hom"))
})

print("NAs counted per column before imputation:")
na_snp_col <- data.frame(na_columnwise(df_snp))
# Handle NAs by imputing the most common value or removing
if (na_mode == "impute") {
  df_snp <- impute_snp(df_snp)
  
} else if (na_mode == "remove") {     # Your removal code here
  df_snp <- omit.NA(df_snp, 10)       # Delete rows with too many NAs
}

print("NAs counted per column after imputation:")
na_columnwise(df_snp)

#df_snp <- df_snp %>%
 #   dplyr::select(-c("rs429358_C"))   #delete columns with mostly NA values (rs429358_C)

#For linear instead of categorical model: change range from 0,1,2 to 0, 0.5, 1
# cols <- colnames(df_snp)[colnames(df_snp) != "eid"]
# df_snp[, (cols) := lapply(.SD, function(x) {
#   levels(x) <- c('0' = '0', '1' = '0.5', '2' = '1')
#   return(x)
# }), .SDcols = cols]

#sanity(df_snp)

#export df_snp
#write.csv(df_snp, file=paste(project_path, "/data/dataframes/df_snp.xlsx")  
#write.xlsx(na_snp_col, file=paste(project_path, "/supplement/NA_count_df_snp.xlsx", sep=''))
```

#### 5. Radiomics

```{r}
# # 7. MRI data df_mri \~50000
# 
# #mri_bulk <- fread("ukb52200.csv", select=c("eid", '20204-2.0','20254-2.0','20203-2.0'))
# #write.csv(mri_bulk, "extracted/mri_bulk.csv", row.names=FALSE)
# 
# setwd(sharepoint_ukb)
# mri_bulk <- fread("extracted/mri_bulk.csv") 
# mri_index <- read_excel("meta/MRI_Abdomen.xlsx")
# setnames(mri_bulk, old = mri_index$datafield1, new = mri_index$Description, skip_absent=TRUE)
# mri_bulk[mri_bulk == ""] <- NA                                #substitutes empty cells for NA
# mri_bulk <- mri_bulk[rowSums(is.na(mri_bulk)) < 3,]         #deletes all rows where all MRI columns are empty
# 
# df_mri <- mri_bulk %>% check_and_remove_withdrawals(df_withdrawals)

```
MRI T1 ShMolli only
```{r}
# setwd(sharepoint_ukb)
# mri_bulk <- fread("extracted/mri_bulk.csv") 
# mri_index <- read_excel("meta/MRI_Abdomen.xlsx")
# setnames(mri_bulk, old = mri_index$datafield1, new = mri_index$Description, skip_absent=TRUE)
# mri_bulk[mri_bulk == ""] <- NA     
# 
# mri_bulk <- mri_bulk %>% select(c("eid", "Liver Imaging - T1 ShMoLLI - DICOM"))
# 
# #substitutes empty cells for NA
# mri_bulk <- mri_bulk[rowSums(is.na(mri_bulk)) < 3,]         #deletes all rows where all MRI columns are empty
# 
# 
# 
# 
# df_mri <- mri_bulk[!is.na(mri_bulk$`Liver Imaging - T1 ShMoLLI - DICOM`) ] %>% check_and_remove_withdrawals(df_withdrawals)

```

# Fusion:

#### Merging (and potential normalization) function + optional removal of features

```{r}
# #This code will apply the merge() function to each pair of data frames in the list list(df_eid, df_covariates, df_diagnosis, df_blood, df_snp), merging them based on the "eid" column.

today <- format(Sys.Date(), "%d_%m_%Y")

# Define overarching function
merge_dataframes <- function(include_metabolomics = FALSE, filter_par = FALSE, normalize = FALSE) {
        
        convert_eid_to_integer <- function(df) { # Function to ensure eid is an integer
          df$eid <- as.integer(df$eid)
          return(df)
        }
      
        # Initial list of dataframes to merge
        dfs_to_merge <- list(df_eid, df_covariates, df_diagnosis, df_blood, df_blood_risk, df_snp)
      
        
        dfs_to_merge <- lapply(dfs_to_merge, convert_eid_to_integer) # Convert 'eid' in each dataframe to integer
      
        
        if (include_metabolomics) { # Conditionally add df_metabolomics and/or else
          dfs_to_merge <- c(dfs_to_merge, list(df_metabolomics))
        }
        
        df_merged <- Reduce(function(x, y) merge(x, y, by = "eid", all = FALSE), dfs_to_merge)  # Merging the dataframes
        
        df_merged <- limit_df(df_merged, mapper = mapper) # applies the function that limits physiologically high and low limits to the df (mostly relevant for covariates). References master_table/Mapper

        # Conditionally apply normalization function
        if (normalize==TRUE) {
          df_merged <- normalize_data(df_merged)
        }
        df_merged <- df_merged %>% select(-all_of(diag_codes)) # Removing the column of interest
        
        #if decided to subset features (e.g. for external validation with less features, put required features in json file and change reduce_model to TRUE for project)
        reduce_model <- if (exists("reduce_model", envir = .GlobalEnv)) get("reduce_model", envir = .GlobalEnv) else FALSE
        vec_remove_columns <- if (exists("vec_remove_columns", envir = .GlobalEnv)) get("vec_remove_columns", envir = .GlobalEnv) else NULL
        
        if (reduce_model) {
          df_merged <- df_merged %>% select(-all_of(vec_remove_columns)) # Removing the columns not necessary
        }
        
        # Filter the "population at risk (par) by a prespecified if required
        if (filter_par) {
          df_merged <- filter_rows_with_pos_entries(df_merged)
        }
        #df_merged <- df_merged %>% select(-all_of(vec_blood_risk)) #remove the blood risk columns again, just used for subsetting, not in the model
        
        # Determine the group status
        if (include_metabolomics) {
            col_subset <<- "met"
        } else {
            col_subset <<- "basic"
        }
        if (filter_par) {
            row_subset <<- "par"
        } else {
            row_subset <<- "all"
        }
        
        if (normalize) {
          assign("raw", "", envir = .GlobalEnv)
        } else {
          assign("raw", "_raw", envir = .GlobalEnv)
        }


        # Remove NAs and return the dataframe
        return(na.omit(df_merged))
      }
```

# Process and save-function: Defining inner/outer layer, merging with y, creating summary report
```{r}
#Initialize the summary report
global_summary <- data.frame(
  Subset = character(),
  Layer = character(),
  Patients = numeric(),
  DOI = numeric(),
  stringsAsFactors = FALSE
)

setwd(project_path)

save_dir <- file.path(project_path, "data", today)
  
  # Check if the directory exists; if not, create it
  if (!dir.exists(save_dir)) {
    dir.create(save_dir, recursive = TRUE)
  }

#Define process and save function
process_and_save <- function(df_x, subset_name) {
  # Generate the desired directory path
  

  # if (row_subset == "par") {
  #         df_x <- filter_rows_with_pos_entries(df_x, vec_risk_constellation)
  # }
          
  for (layer in c("INNER", "OUTER")) {
    df_x_temp <- df_x  # Work with a copy of df_x to avoid modifying the original df_x

    if (layer == "INNER") {
      df_y <- read.csv(paste(project_path, "/data/dataframes/df_y.csv", sep='')) %>%
        subset(split_ext == 0) %>%
        select(c("eid", "split_int", "status", "status_cancerreg", "date_of_diag", "assessment", "difftime"))
      #df_y$eid <- as.integer(df$eid)
      print("Inner layer:")
    } 
    
    else {
      df_y <- read.csv(paste(project_path, "/data/dataframes/df_y.csv", sep='')) %>%
        select(c("eid", "split_ext", "status", "status_cancerreg", "date_of_diag", "assessment", "difftime")) %>%
        subset(split_ext == 1)
      #df_y$eid <- as.integer(df$eid)
      cat("\n")
      print("Outer layer:")
    }
    
    df_temp <- df_x_temp %>%
        inner_join(df_y, by = "eid")
    nr_doi <- sum(df_temp$status)             #Amount of positive cases
    nr_all <- length(unique(df_temp$eid))       #All rows  
    
    # Update the global dataframe
    new_entries <- data.frame(
      Subset = subset_name,
      Layer = layer,
      Patients = nr_all,
      Patients_with_DOI = nr_doi,
      stringsAsFactors = FALSE
    )
    global_summary <<- rbind(global_summary, new_entries)
    print(paste("Patients in", row_subset, "subset:", nr_all))
    print(paste("Thereof patients with ", DOI, ": ", nr_doi))
    df_x_temp <- df_temp %>%                  # Select all BUT the status (= info to be predicted)
      select(-(c("status", "status_cancerreg")))      #select(-(c("status", "Date of assessment")))
    df_y_temp <- df_temp %>%                  # Select JUST the status
      select(c("eid", "status", "status_cancerreg", "date_of_diag", "assessment", "difftime"))  
    
    write.csv(df_x_temp, file=paste(project_path, "/data/", today, "/X_", tolower(layer), "_", col_subset, "_", subset_name, raw, ".csv", sep=""), row.names=FALSE)
    write.csv(df_y_temp, file=paste(project_path, "/data/", today, "/y_", tolower(layer), "_", col_subset, "_", subset_name, raw, ".csv", sep=""), row.names=FALSE)
    
    
    
    ### To be implemented: Adapt export to RDtata format: (Works, but implementation in Python not yet)
    
    # save(df_x_temp, file=paste0(project_path, "/data/", today, "/X_", tolower(layer), "_", col_subset, "_", subset_name, "_", raw, ".RData"))
    # save(df_y_temp, file=paste0(project_path, "/data/", today, "/y_", tolower(layer), "_", col_subset, "_", subset_name, "_", raw, ".RData"))
    
    # Assign the dataframes to global variables for manual inspection
    #assign(paste0("df_X_check", row_subset, col_subset, layer), df_x_temp, envir = .GlobalEnv)
    #assign(paste0("df_y_check", row_subset, col_subset, layer), df_y_temp, envir = .GlobalEnv)
  }
}

```


Execution of normalized merge/process_and save (always execute both rows! If not, the global variable "row_subset" will not be set correctly)

### Here, we have to merge X and y for once, to make sure that 
a) for X, all patients with DOI before ukb visit are excluded
b) for y, that all patients with missing data are excluded from y as well
c) in the end, both tables have the same length


### Normalization function
```{r}
vec_covariates <- c("MultipleDeprivationIndex", "Pack years", "Waist circumference", # Columns for "simple" Minmax in covariates
                  "Weight", "Standing height", "Alk_g_d", "Bloodpressure_sys", "Bloodpressure_dia", "BMI", "AGE", "Handgripstrength")
vec_pc <- c("PC1", "PC2", "PC3", "PC4", "PC5")
#vec_blood / vec_metabolomics are already defined above
vec_all <- unique(c(vec_covariates, vec_pc, vec_blood, vec_metabolomics)) 
  

# Function to normalize data (deployed later inside merge function)
normalize_data <- function(df, df_cap = NULL, cap = FALSE) {
  

  # Adjust PC columns to get positive values
  for (i in vec_pc) { 
    df[[i]] <- df[[i]] + abs(min(df[[i]], na.rm = TRUE))
  }
  
  # Apply minmax normalization to remaining or all columns
  for (i in vec_all) {
    if (i != "eid" && i %in% names(df) && is.numeric(df[[i]])) {
      df[[i]] <- minmax(df[[i]])
    }
  }
  
  return(df)
}
```

# Export normalized data
```{r}

export_normalized <- TRUE

par_subset <- project_configs[[project_key]]$par_subset #load project-specific subsets of patients-at-risk

if (export_normalized) {
  
  # For all data (without filtering by chronic liver disease and without metabolomics)
  df_x_all <- merge_dataframes(normalize=TRUE)
  process_and_save(df_x_all, "all")
  
  # For data including metabolomics
  df_x_all_met <- merge_dataframes(include_metabolomics = TRUE, normalize=TRUE)
  process_and_save(df_x_all_met, "all")
  
  # For data filtered by chronic liver disease
  df_x_par <- merge_dataframes(filter_par = TRUE, normalize=TRUE)
  process_and_save(df_x_par, "par")
  write.csv(df_x_par$eid, file = file.path(project_path, "data", today, "par_eids.csv"), row.names = FALSE)
  
  # For data filtered by chronic liver  + metabolomics
  df_x_par_met <- merge_dataframes(filter_par = TRUE, include_metabolomics=TRUE, normalize=TRUE)
  process_and_save(df_x_par_met, "par")
  
  
  
  # # For additional subsets
  # 
  # par_subset <- "CLD"
  # df_x_par <- merge_dataframes(filter_par = TRUE, normalize=TRUE)
  # process_and_save(df_x_par, "par_CLD")
  # write.csv(df_x_par$eid, file = file.path(project_path, "data", today, "par_CLD_eids.csv"), row.names = FALSE)

}

par_subset
  


```
# Export absolute data / Create original dataframe X for Table 1 of Publication (Overview of patient data)
```{r}
export_absolute <- TRUE

if (export_absolute) {

# For all data (without filtering by chronic liver disease and without metabolomics)
df_x_all_raw <- merge_dataframes(normalize=FALSE)
process_and_save(df_x_all_raw, "all")

# For data including metabolomics
df_x_all_met_raw <- merge_dataframes(include_metabolomics = TRUE, normalize=FALSE)
process_and_save(df_x_all_met_raw, "all_met")

# For data filtered by chronic liver disease
df_x_par_raw <- merge_dataframes(filter_par = TRUE, normalize=FALSE)
process_and_save(df_x_par_raw, "par")

# For data filtered by chronic liver  + metabolomics
df_x_par_met_raw <- merge_dataframes(filter_par = TRUE, include_metabolomics=TRUE, normalize=FALSE)
process_and_save(df_x_par_met_raw, "par_met")
}
```


# Summary report (Excel file with count per layer/status, and metadata)
```{r}
#Summary report
grouped_summary <- global_summary %>%
  group_by(Subset) %>%
  summarise(
    Total_Patients = sum(Patients),
    Total_Patients_with_DOI = sum(Patients_with_DOI)
  )
df_risk_constellation <- data.frame(Risk_constellation = vec_risk_constellation)
df_diags <- data.frame(Integrated_ICD = IOIs)

sheets_list <- list(
  "Per_Layer_Summary" = global_summary,
  "All_Summary" = grouped_summary,
  "Risk Constellation" = df_risk_constellation,
  "Integrated Diagnosis" = df_diags
)
write_xlsx(sheets_list, path = (paste(project_path, "/data/", today, "/summary_reports.xlsx", sep = "")))
```

# Create columngroups table (creates table with column names and to which initial df they belong (necessary for feature importance later))
```{r}
# Create a list of the dataframes and their names
dfs = list(df_eid=df_eid, df_covariates=df_covariates, df_diagnosis=df_diagnosis, df_blood=df_blood, df_snp=df_snp, df_metabolomics=df_metabolomics)

# Use lapply to loop over the dataframes and create the temporary dataframes
temp_dfs = lapply(names(dfs), function(df_name) {
  data.frame(column_name = colnames(dfs[[df_name]]), source_df = df_name)
})

# Combine the temporary dataframes into one dataframe
df_columngroups = do.call(rbind, temp_dfs) %>%
  subset(column_name != "eid") %>%
  rbind(data.frame(column_name = "split_int", source_df = "df_metadata")) %>%
  rbind(data.frame(column_name = "split_ext", source_df = "df_metadata")) %>%
  mutate(source_df = ifelse(startsWith(column_name, "PC"), "df_snp", source_df)) #Manually change columngroups relationship as needed

df_columngroups <- df_columngroups[!df_columngroups$column_name %in% vec_remove_columns, ]

today <- format(Sys.Date(), "%d_%m_%Y")
write.csv(df_columngroups, paste(project_path, "/data/", today, "/columngroups.csv", sep=""), row.names=FALSE)

rm(dfs) #Inefficient way for creating columngroups, uses lots of memory


```




# Table Export Preparation

Prepare and modify the df_x_all_raw df to be more visually appealing
when creating the table 1 for the paper

```{r}
setwd(project_path)
load("data/dataframes/df_y.RData") 
df_y <- df_y %>% select(c("eid", "status"))

covariates_list <- df_columngroups[df_columngroups$source_df == "df_covariates", "column_name"]
icd_list <- df_columngroups[df_columngroups$source_df == "df_diagnosis", "column_name"]
metabolomics_list <- df_columngroups[df_columngroups$source_df == "df_metabolomics", "column_name"]
blood_list <- df_columngroups[df_columngroups$source_df == "df_blood", "column_name"]
snp_list <- df_columngroups[df_columngroups$source_df == "df_snp", "column_name"]


#List of label strings: Name of Column 
label_list <- list(
  AGE = "Age [years]",
  location_name = "Location",
  location_country = "Country",
  Family_diabetes = "Family Diabetes",
  DM = "Diabetes mellitus",
  Alk_g_d = "Alcohol [g/d]",
  High_Alk = "High Alcohol Consumption",
  Path_Alk = "Pathological Alcohol Consumption",
  BMI_cat = "BMI Categories",
  Bloodpressure_sys = "Bloodpressure sys. [mmHg]",
  Weight = "Weight [kg]",
  'Standing height' = "Standing height [cm]",
  'Waist circumference' = "Waist circumference [cm]"
)

list(age ~ "Age, years", response ~ "Tumor Response")

table1_order <- c(
  "AGE", "SEX", "BMI", "BMI_cat", "Waist circumference", 
  "Weight", "Standing height", "Ethnicity", 
  "MultipleDeprivationIndex", "Bloodpressure_sys", "Medication", "DM",
  "Family_diabetes", "Pack years", "Alk_g_d" # Continue as necessary
)



table1b_order <- c(
  "AGE", "SEX", "Ethnicity", "BMI", "Waist circumference", 
  "Weight", "Standing height",
  "MultipleDeprivationIndex", "Bloodpressure_sys", "Medication", "Pack years" # Continue as necessary
)

table_alc_order <- c("Alc")




```

# Table Creation All

```{r}
head_only <- FALSE
# Change "all" to "par" if you want to create tables of just the "patient at risk subset
df_all <- df_x_all_raw %>% inner_join(df_y, by = "eid")
df_all$status <- ifelse(df_all$status == 0, paste("No ", DOI), 
                                  ifelse(df_all$status == 1, DOI, df_all$status))

df_tbl_1 <- df_all %>%
  select(!any_of(c(blood_list, metabolomics_list, 
                   setdiff(icd_list, "DM"),  # Remove all icd_list items except "DM"
                   "eid", "Date of assessment", "UKB assessment centre", 
                   snp_list)))

create_table(df_tbl_1, "Table 1", export_RDS=TRUE, head_only=head_only, remove_SEX=FALSE,  enforced_order=table1_order)


df_tbl_1b <- df_all %>%
  select(!any_of(c(blood_list, metabolomics_list, icd_list, "eid", "Date of assessment", "UKB assessment centre", snp_list)))

create_table(df_tbl_1b, "Table 1b", export_RDS=TRUE, head_only=head_only, remove_SEX=FALSE,  enforced_order=table1b_order)
```

# Table Creation PAR
```{r}
head_only <- FALSE
df_all <- df_x_par_raw %>% inner_join(df_y, by = "eid")
df_all$status <- ifelse(df_all$status == 0, paste("No", DOI), 
                                  ifelse(df_all$status == 1, DOI, df_all$status))

df_tbl_1 <- df_all %>%
  select(!any_of(c(blood_list, metabolomics_list, icd_list, "eid", "Date of assessment", "UKB assessment centre", snp_list)))

create_table(df_tbl_1, "Table 1_par", export_RDS=TRUE, head_only=head_only, remove_SEX=FALSE,  enforced_order=table1_order)
```

# Tables ICD/Blood
```{r}
head_only <- FALSE
df_all <- df_x_all_raw %>% inner_join(df_y, by = "eid")
df_all$status <- ifelse(df_all$status == 0, paste("No ", DOI), 
                                  ifelse(df_all$status == 1, DOI, df_all$status))


df_tbl_icd <- df_all %>%
  select(any_of(c("status", "SEX", icd_list)))

df_tbl_blood <- df_all %>%
  select(any_of(c("status", "SEX", blood_list)))



df_tbl_snp <- df_all %>%
  select(any_of(c("status", "SEX", snp_list)))


df_tbl_metabolomics <- df_x_all_met_raw %>% inner_join(df_y, by = "eid") %>%
  mutate(status = if_else(status == 1, DOI, paste0("No ", DOI))) %>%
  select(any_of(c("status", "SEX", metabolomics_list)))

  
head_only <- FALSE

create_table(df_tbl_blood, "Table Blood", export_RDS=TRUE, head_only=head_only)

create_table(df_tbl_icd, "Table ICD", export_RDS=TRUE, head_only=head_only, create_binary_table = TRUE)

create_table(df_tbl_metabolomics, "Table Metabolomics", export_RDS=FALSE, head_only=head_only)

create_table(df_tbl_snp, "Table SNP", export_RDS=TRUE, head_only=head_only)



```



# Male/Female stratified table
```{r}
head_only <- FALSE
# Change "all" to "par" if you want to create tables of just the "patient at risk subset
df_all <- df_x_all_raw %>% inner_join(df_y, by = "eid")
df_all$status <- ifelse(df_all$status == 0, paste("No ", DOI), 
                                  ifelse(df_all$status == 1, DOI, df_all$status))


df_tbl_1 <- df_all %>%
  select(!any_of(c(blood_list, metabolomics_list, 
                   setdiff(icd_list, "DM"),  # Remove all icd_list items except "DM"
                   "eid", "Date of assessment", "UKB assessment centre", 
                   snp_list)))


split_create_merge_tables(df_tbl_1, table_name="Table1", feature="SEX", enforced_order=table1_order, remove_SEX=TRUE, export_RDS=TRUE)

Table_1_stratified <- import_merge_tables(table_name= "Table1", feature="SEX", levels = c("Female", "Male"))



#Shorter Table 1


# df_tbl_1b <- df_all %>%
#   select(!any_of(c(blood_list, metabolomics_list, icd_list, "eid", "Date of assessment", "UKB assessment centre", snp_list)))
# 
# 
# split_create_merge_tables(df_tbl_1b, feature="SEX", table_name_prefix="SEX", enforced_order=table1b_order, remove_SEX=FALSE, export_RDS=TRUE)
# 
# Table_1b_stratified <- import_merge_tables(table_name= "Table_1b", feature="SEX", levels = c("Female", "Male"))



# Stratified "Blood Table"
split_create_merge_tables(df_tbl_blood, table_name= "Table_Blood", feature="SEX", enforced_order=FALSE, remove_SEX=TRUE, export_RDS=TRUE)

Table_Blood_stratified <- import_merge_tables(table_name= "Table_Blood", feature="SEX", levels = c("Female", "Male"))

# Stratified "ICD Table"
split_create_merge_tables(df_tbl_icd, table_name= "Table_ICD", feature="SEX", enforced_order=FALSE, remove_SEX=TRUE, export_RDS=TRUE, create_binary_table = TRUE)

Table_ICD_stratified <- import_merge_tables(table_name= "Table_ICD", feature="SEX", levels = c("Female", "Male"))


```



































#Check cause of missing data (merge df_y with every single df to check how many cases are lost)
Does not work yet
```{r}
setwd(project_path)
load("data/dataframes/df_y.RData")
inspect_na_contributions <- function(df_y, include_metabolomics = FALSE) {
  
  # Initial list of dataframes to merge
  dfs_to_merge <- list(df_covariates, df_diagnosis, df_blood, df_snp) # Removing df_eid for the analysis
  
  # Optionally add df_metabolomics
  if (include_metabolomics) {
    dfs_to_merge <- c(dfs_to_merge, list(df_metabolomics))
  }
  
  # Function to calculate number of rows dropped due to NA after merging with df_y
  dropped_rows_due_to_na <- function(df) {
    merged_df <- merge(df_y, df, by = "eid", all = TRUE)
    return(sum(!is.na(merged_df$status) & apply(merged_df[-1], 1, function(row) all(is.na(row)))))
  }
  
  # Apply function to each dataframe
  na_drop_counts <- sapply(dfs_to_merge, dropped_rows_due_to_na)
  
  # Assign names based on the condition of include_metabolomics
  if (include_metabolomics) {
    names(na_drop_counts) <- c("covariates", "icd", "blood", "snp", "metabolomics")
  } else {
    names(na_drop_counts) <- c("covariates", "icd", "blood", "snp")
  }
  
  return(na_drop_counts)
}

na_contributions <- inspect_na_contributions(df_y, include_metabolomics = TRUE)
print(na_contributions)



```




